File: scripts/library.js
// La bibliothèque unifiée : agrège tous les providers de plateforme.
//
// Aujourd'hui seul Steam est implémenté, mais la couture existe dès maintenant :
// chaque jeu porte SA plateforme d'origine, et ajouter Epic / EA / GOG / Ubisoft
// revient à déposer un fichier dans platforms/ et à l'ajouter à la liste.
//
// Contrat d'un provider :
//   id       : identifiant court, stable ('steam')
//   name     : nom affichable ('Steam')
//   detect() : bool — la plateforme est-elle présente sur la machine ?
//   scan()   : { games: [...] } — chaque jeu porte platform, id, name, launch, art
//
// Un provider qui échoue ne doit JAMAIS faire tomber les autres : une bibliothèque
// partielle vaut mieux qu'un écran vide.

const steam = require('./platforms/steam');

// À venir : epic, ea, ubisoft, gog. Voir docs/roadmap.md.
const PROVIDERS = [steam];

/** Métadonnées d'affichage par plateforme. Le badge est volontairement court. */
const PLATFORMS = {
  steam:   { name: 'Steam',           badge: 'STEAM' },
  epic:    { name: 'Epic Games',      badge: 'EPIC' },
  ea:      { name: 'EA App',          badge: 'EA' },
  ubisoft: { name: 'Ubisoft Connect', badge: 'UBI' },
  gog:     { name: 'GOG',             badge: 'GOG' },
  xbox:    { name: 'Xbox / Game Pass', badge: 'XBOX' },
};

function scanAll() {
  const games = [];
  const platforms = [];

  for (const p of PROVIDERS) {
    if (!p.detect()) {
      platforms.push({ id: p.id, name: p.name, present: false, count: 0 });
      continue;
    }
    try {
      const res = p.scan();
      games.push(...res.games);
      platforms.push({ id: p.id, name: p.name, present: true, count: res.games.length });
    } catch (e) {
      // Un provider cassé est signalé, pas fatal.
      platforms.push({ id: p.id, name: p.name, present: true, count: 0, error: e.message });
    }
  }

  // Tri par récence, toutes plateformes confondues — loi n°4.
  games.sort((a, b) => b.lastPlayed - a.lastPlayed);
  return { games, platforms };
}

/** Retrouve un jeu par plateforme + id. Utilisé pour servir les jaquettes. */
function findGame(games, platform, id) {
  return games.find(g => g.platform === platform && String(g.id) === String(id)) || null;
}

// --- Sortie lisible : node scripts/library.js ---
if (require.main === module) {
  const { games, platforms } = scanAll();

  for (const p of platforms) {
    const state = !p.present ? 'absente' : p.error ? `ERREUR: ${p.error}` : `${p.count} jeux`;
    console.log(`${p.name.padEnd(18)} ${state}`);
  }
  console.log(`\n${games.length} jeux au total\n`);

  const size = n => (n ? (n / 1e9).toFixed(1) + ' Go' : '—');
  const date = t => (t ? new Date(t * 1000).toLocaleDateString('fr-FR') : '—');

  for (const g of games) {
    const art = ['portrait', 'hero', 'logo'].filter(k => g.art[k]).join('+') || 'AUCUN VISUEL';
    console.log(
      `${(PLATFORMS[g.platform]?.badge || g.platform).padEnd(6)} ${String(g.id).padStart(8)}  ` +
      `${g.name.slice(0, 34).padEnd(34)} ${size(g.sizeBytes).padStart(8)}  ${date(g.lastPlayed).padStart(10)}  ${art}`
    );
  }
}

module.exports = { scanAll, findGame, PLATFORMS };

File: src-tauri/src/library.rs
//! La bibliothèque unifiée : agrège tous les providers de plateforme.
//! Miroir Rust de scripts/library.js. Aujourd'hui seul Steam est câblé,
//! mais chaque jeu porte déjà sa plateforme d'origine — ajouter un provider
//! revient à déposer un module dans platforms/ et à l'ajouter à scan_all().

use crate::platforms::{custom, steam};
use serde::Serialize;

#[derive(Serialize, Clone, Default)]
pub struct Art {
    pub portrait: Option<String>,
    pub hero: Option<String>,
    pub logo: Option<String>,
}

/// Comment démarrer un jeu — deux mécaniques, pas une.
///
/// `Uri` : un protocole résolu par l'OS (steam://, uplay://, origin2://…).
/// `Exec` : un exécutable lancé directement avec ses arguments — le cas des
/// émulateurs (`pcsx2.exe "rom.chd"`) et des .exe ajoutés à la main via le
/// bouton +. Poser les deux formes maintenant évite de refaire ce refactor
/// à travers tout le code quand l'émulation ou le bouton + arriveront.
#[derive(Clone, Debug)]
pub enum Launch {
    Uri(String),
    // Pas encore construite : arrive avec le bouton + et l'émulation
    // (chantiers suivants). Posée maintenant pour que launch_game() gère
    // déjà les deux mécaniques sans refactor à venir.
    #[allow(dead_code)]
    Exec { path: String, args: Vec<String> },
}

#[derive(Serialize, Clone)]
pub struct Game {
    pub platform: String,
    pub id: String,
    pub name: String,
    // Jamais envoyé au renderer : exposition minimale, et de toute façon
    // inutile côté client — c'est launch_game() qui l'exécute, côté natif.
    #[serde(skip)]
    pub launch: Launch,
    #[serde(rename = "sizeBytes")]
    pub size_bytes: u64,
    #[serde(rename = "lastPlayed")]
    pub last_played: i64,
    pub art: Art,
}

#[derive(Serialize)]
pub struct PlatformStatus {
    pub id: String,
    pub name: String,
    pub present: bool,
    pub count: usize,
    pub error: Option<String>,
}

#[derive(Serialize)]
pub struct ScanResult {
    pub games: Vec<Game>,
    pub platforms: Vec<PlatformStatus>,
}

/// Fait tourner un provider et range son résultat — factorisé pour ne pas
/// recopier ce bloc à chaque plateforme ajoutée (Epic, EA, GOG suivront
/// le même moule).
fn run_provider(
    id: &str,
    name: &str,
    present: bool,
    scan: impl FnOnce() -> Result<Vec<Game>, String>,
    games: &mut Vec<Game>,
    platforms: &mut Vec<PlatformStatus>,
) {
    if !present {
        platforms.push(PlatformStatus { id: id.into(), name: name.into(), present: false, count: 0, error: None });
        return;
    }
    match scan() {
        Ok(g) => {
            platforms.push(PlatformStatus { id: id.into(), name: name.into(), present: true, count: g.len(), error: None });
            games.extend(g);
        }
        Err(e) => platforms.push(PlatformStatus { id: id.into(), name: name.into(), present: true, count: 0, error: Some(e) }),
    }
}

/// Un provider en échec est signalé, jamais fatal aux autres :
/// une bibliothèque partielle vaut mieux qu'un écran vide.
pub fn scan_all() -> ScanResult {
    let mut games = Vec::new();
    let mut platforms = Vec::new();

    run_provider(steam::ID, steam::NAME, steam::detect(), steam::scan, &mut games, &mut platforms);
    run_provider(custom::ID, custom::NAME, custom::detect(), custom::scan, &mut games, &mut platforms);

    // Tri par récence, toutes plateformes confondues — loi n°4.
    games.sort_by(|a, b| b.last_played.cmp(&a.last_played));
    ScanResult { games, platforms }
}

pub fn find_game<'a>(games: &'a [Game], platform: &str, id: &str) -> Option<&'a Game> {
    games.iter().find(|g| g.platform == platform && g.id == id)
}
